std/selector/multiplexer.go

```diff
@@ -50,15 +53,49 @@ func Map(api frontend.API, queryKey frontend.Variable,
 // Mux is an n to 1 multiplexer: out = inputs[sel]. In other words, it selects
 // exactly one of its inputs based on sel. The index of inputs starts from zero.
 //
-// sel needs to be between 0 and n - 1 (inclusive), where n is the number of
-// inputs, otherwise the proof will fail.
+// sel needs to be between 0 and n - 1 (inclusive), where n is the number of the
+// inputs, otherwise the proof will fail or silently return another element from the inputs.
 func Mux(api frontend.API, sel frontend.Variable, inputs ...frontend.Variable) frontend.Variable {
-	// we use BinaryMux when len(inputs) is a power of 2.
-	if binary.OnesCount(uint(len(inputs))) == 1 {
-		selBits := bits.ToBinary(api, sel, bits.WithNbDigits(binary.Len(uint(len(inputs)))-1))
+	n := uint(len(inputs))
+	if n == 1 {
+		return inputs[0]
+	}
+
+	nbBits := binary.Len(n - 1)
+	selBits := bits.ToBinary(api, sel, bits.WithNbDigits(nbBits))
+
+	// We use BinaryMux when len(inputs) is a power of 2.
+	if binary.OnesCount(n) == 1 {
 		return BinaryMux(api, selBits, inputs)
 	}
-	return dotProduct(api, inputs, Decoder(api, len(inputs), sel))
+
+	// If sel is beyond range [0, n-1], bcmp might wrongly produce reversed results,
+	// leading Mux to return another elements from the inputs.
+	// See doc of cmp.NewBoundedComparator.
+	bcmp := cmp.NewBoundedComparator(api, big.NewInt(int64(n)), false)
```

Comment:
potential soundness issue here if `sel >= n`

Reply:
It is still desirable when the value range of `sel` is already constrained. Let's add it as a new function instead of replacing the existing `Mux`. We need a proper name so that its limitation is apparent.